src/contracts/core/Slasher.sol

```diff
@@ -1,102 +1,368 @@
+     * @param strategies the list of strategies slashing requested is being reduced for
+     * @param epoch the epoch in which slashing was requested
+     * @param bipsToReduce the basis points slashing to reduced for given strategies
+     */
+    function reduceRequestedBipsToSlash(
+        address operator,
+        bytes4 operatorSetID,
+        IStrategy[] memory strategies,
+        uint32 epoch,
+        uint32 bipsToReduce
+    ) external {
+        uint32 currentEpoch = EpochUtils.currentEpoch();
+        require(
+            epoch == currentEpoch || epoch + 1 == currentEpoch,
+            "Slasher.reduceRequestedBipsToSlash: can only reduce for current or previous epoch"
+        );
+        require(bipsToReduce > 0, "Slasher.reduceRequestedBipsToSlash: bipsToReduce must be negative");
+        require(
+            bipsToReduce < 2 ** 31, "Slasher.reduceRequestedBipsToSlash: bipsToReduce must be less than minimum int32"
+        );
+        _modifyRequestedBipsToSlash(operator, operatorSetID, strategies, epoch, -int32(bipsToReduce));
+    }
 
-    function freezeOperator(address) external {}
+    /**
+     * @notice Permissionlessly called to execute slashing of a given list of
+     * strategies for a given operator, for the latest unslashed epoch
+     *
+     * @param operator the operator to slash
+     * @param strategies the list of strategies to execute slashing for
+     * @param epoch the epoch in which the slashing requests to execute were made
+     */
+    function executeSlashing(address operator, IStrategy[] memory strategies, uint32 epoch) external {
+        require(
+            EpochUtils.currentEpoch() > EpochUtils.getMinExecutionEpochFromRequestEpoch(epoch),
+            "Slasher.executeSlashing: current epoch must be greater than the minimum execution epoch"
+        );
+        for (uint256 i = 0; i < strategies.length; ++i) {
+            IStrategy strategy = strategies[i];
+            // make sure the slashing request is in order and overwrite the last executed slashing request ID
+            SlashingRequest memory slashingRequest = slashingRequests[operator][strategy][epoch];
+            require(
+                slashingRequestIds[operator][strategy].lastExecutedSlashingRequestId + 1 == slashingRequest.id,
+                "Slasher.executeSlashing: must execute slashings in order"
+            );
+            slashingRequestIds[operator][strategy].lastExecutedSlashingRequestId = slashingRequest.id;
 
-    function resetFrozenStatus(address[] calldata) external {}
+            // truncate to BIPS_FACTOR_SQUARED if it exceeds it
+            if (slashingRequest.slashingRate > SlashingAccountingUtils.BIPS_FACTOR_SQUARED) {
+                slashingRequest.slashingRate = SlashingAccountingUtils.BIPS_FACTOR_SQUARED;
+            } else if (slashingRequest.slashingRate == 0) {
+                // if the slashing rate is 0 due to cancellation, skip the slashing
+                continue;
+            }
```

Comment:
I think we should revert instead of continuing in this case?
there is no race condition; is there a reason to prefer this behavior?